src/wasm.h

```diff
@@ -1233,13 +1234,104 @@ std::ostream& Expression::print(std::ostream &o, unsigned indent) {
   return o;
 }
 
+template<typename SubType, typename ReturnType>
+struct WasmWalkerBase : public WasmVisitor<SubType, ReturnType> {
+  virtual void walk(Expression*& curr) { abort(); }
+  virtual void startWalk(Function *func) { abort(); }
+  virtual void startWalk(Module *module) { abort(); }
+};
+
+template<typename ParentType>
+struct ChildWalker : public WasmWalkerBase<ChildWalker<ParentType>, void> {
+  ParentType& parent;
+
+  ChildWalker(ParentType& parent) : parent(parent) {}
+
+  void visitBlock(Block *curr) {
+    ExpressionList& list = curr->list;
+    for (size_t z = 0; z < list.size(); z++) {
```

Comment:
Is this kind of optimization still worth doing in C++?

Also, I tried using a range-based `for` loop, but the tests no longer pass. @kripken do we mutate these vectors while iterating over them?